### PR TITLE
Deprecate CommonTermsQuery and cutoff_frequency

### PR DESCRIPTION
**Pull request by matriv**
Since the max_score optimization landed in Elasticsearch 7,
the `CommonTermsQuery` is redundant and slower. Moreover the
`cutoff_frequency` parameter for `MatchQuery` and `MultiMatchQuery`
is also redundant.

Relates to #37096

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java [hunk 1]
```diff
@@ -63,7 +72,12 @@
     private static final ParseField SLOP_FIELD = new ParseField("slop");
     private static final ParseField ZERO_TERMS_QUERY_FIELD = new ParseField("zero_terms_query");
     private static final ParseField LENIENT_FIELD = new ParseField("lenient");
-    private static final ParseField CUTOFF_FREQUENCY_FIELD = new ParseField("cutoff_frequency");
+    /**
+     * @deprecated Since max_score optimization landed in 7.0, normal MultiMatchQuery
+     *             will achieve the same result without any configuration.
+     */
+    @Deprecated
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java]: This is not needed, the deprecated annotation is for java user that uses MultiMatchQueryBuilder#cutoffFrequency in their code.

File: server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java [hunk 2]
```diff
@@ -91,6 +105,10 @@
     private String fuzzyRewrite = null;
     private Float tieBreaker;
     private Boolean lenient;
+    /**
+     * @deprecated See {@link MultiMatchQueryBuilder#CUTOFF_FREQUENCY_FIELD} for more details
+     */
+    @Deprecated
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java]: The annotation on the setter is enough, no need to deprecate the variable

File: server/src/main/java/org/elasticsearch/index/search/MatchQuery.java [hunk 3]
```diff
@@ -171,6 +172,10 @@ public void writeTo(StreamOutput out) throws IOException {
 
     protected ZeroTermsQuery zeroTermsQuery = DEFAULT_ZERO_TERMS_QUERY;
 
+    /**
+     * @deprecated See {@link MatchQueryBuilder#setCommonTermsCutoff(Float)} for more details
+     */
+    @Deprecated
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/search/MatchQuery.java]: Same here, this is not needed

File: server/src/main/java/org/elasticsearch/index/search/MatchQuery.java [hunk 4]
```diff
@@ -298,6 +307,10 @@ protected final Query parseInternal(Type type, String fieldName, MatchQueryBuild
         return query == null ? zeroTermsQuery() : query;
     }
 
+    /**
+     * @deprecated See {@link CommonTermsQueryBuilder}
+     */
+    @Deprecated
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/search/MatchQuery.java]: This is internal (private) so no need to add the annotation

File: server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryBuilderTests.java [hunk 5]
```diff
@@ -37,8 +38,19 @@
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.nullValue;
 
+/**
+ * @deprecated See {@link CommonTermsQueryBuilder}
+ */
+@Deprecated
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryBuilderTests.java]: Not needed

File: server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryParserTests.java [hunk 6]
```diff
@@ -22,10 +22,12 @@
 import org.elasticsearch.action.search.SearchResponse;
 import org.elasticsearch.test.ESSingleNodeTestCase;
 
-import java.io.IOException;
-
+/**
+ * @deprecated See {@link CommonTermsQueryBuilder}
+ */
+@Deprecated
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryParserTests.java]: This is not needed

File: server/src/test/java/org/elasticsearch/search/profile/query/RandomQueryGenerator.java [hunk 7]
```diff
@@ -169,6 +169,10 @@ private static QueryBuilder randomConstantScoreQuery(List<String> stringFields,
         return QueryBuilders.constantScoreQuery(randomQueryBuilder(stringFields, numericFields, numDocs, depth - 1));
     }
 
+    /**
+     * @deprecated See {@link CommonTermsQueryBuilder}
+     */
+    @Deprecated
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/profile/query/RandomQueryGenerator.java]: This is not needed (the deprecated annotation) and since it will generate a deprecation warning I wonder if this function should be removed entirely. Random queries that generates warning are difficult to handle in tests so I am leaning towards removing.

**Review by jimczi — COMMENTED** [review]
Thanks @matriv, I left some comments but it looks good overall

**Comment by matriv** [comment]
@jimczi Thanks for the review! Comments are addressed.

File: server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java [hunk 8]
```diff
@@ -63,7 +72,7 @@
     private static final ParseField SLOP_FIELD = new ParseField("slop");
     private static final ParseField ZERO_TERMS_QUERY_FIELD = new ParseField("zero_terms_query");
     private static final ParseField LENIENT_FIELD = new ParseField("lenient");
-    private static final ParseField CUTOFF_FREQUENCY_FIELD = new ParseField("cutoff_frequency");
+    private static final ParseField CUTOFF_FREQUENCY_FIELD = new ParseField("cutoff_frequency", "cutoff_frequency");
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java]: I think you can replace this with  `new ParseField("cutoff_frequency") .withAllDeprecated("CUTOFF_FREQUENCY_DEPRECATION_MSG")`

File: server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java [hunk 9]
```diff
@@ -484,8 +493,13 @@ public boolean lenient() {
      * Set a cutoff value in [0..1] (or absolute number &gt;=1) representing the
      * maximum threshold of a terms document frequency to be considered a low
      * frequency term.
+     *
+     * @deprecated Since max_score optimization landed in 7.0, normal MultiMatchQuery
+     *             will achieve the same result without any configuration.
      */
+    @Deprecated
     public MultiMatchQueryBuilder cutoffFrequency(float cutoff) {
+        deprecationLogger.deprecated(CUTOFF_FREQUENCY_DEPRECATION_MSG);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java]: I don't think you need to use the deprecation logger here. The Deprecated annotation should be enough.

File: server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java [hunk 10]
```diff
@@ -494,8 +508,15 @@ public MultiMatchQueryBuilder cutoffFrequency(float cutoff) {
...
      * frequency term.
+     *
+     * @deprecated Since max_score optimization landed in 7.0, normal MultiMatchQuery
+     *             will achieve the same result without any configuration.
      */
+    @Deprecated
     public MultiMatchQueryBuilder cutoffFrequency(Float cutoff) {
+        if (cutoff != null) {
+            deprecationLogger.deprecated(CUTOFF_FREQUENCY_DEPRECATION_MSG);
+        }
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java]: Same here, you don't need to log the deprecation

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 11]
```diff
@@ -235,8 +249,12 @@ public int maxExpansions() {
      * Set a cutoff value in [0..1] (or absolute number &gt;=1) representing the
      * maximum threshold of a terms document frequency to be considered a low
      * frequency term.
+     *
+     * @deprecated see {@link MatchQueryBuilder#CUTOFF_FREQUENCY_FIELD} for more details
      */
+    @Deprecated
     public MatchQueryBuilder cutoffFrequency(float cutoff) {
+        deprecationLogger.deprecated(CUTOFF_FREQUENCY_DEPRECATION_MSG);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java]: You don't need to log the deprecation

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java [hunk 12]
```diff
@@ -42,8 +44,20 @@
...
+        "[max_score] optimization which is applied automatically without any configuration";
+
     public static final ParseField ZERO_TERMS_QUERY_FIELD = new ParseField("zero_terms_query");
-    public static final ParseField CUTOFF_FREQUENCY_FIELD = new ParseField("cutoff_frequency");
+    /**
+     * @deprecated Since max_optimization optimization landed in 7.0, normal MatchQuery
+     *             will achieve the same result without any configuration.
+     */
+    @Deprecated
+    public static final ParseField CUTOFF_FREQUENCY_FIELD = new ParseField("cutoff_frequency", "cutoff_frequency");
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java]: You can  remove the Deprecated annotation and use:
````
new ParseField("cutoff_frequency") .withAllDeprecated(CUTOFF_FREQUENCY_DEPRECATION_MSG)
````

File: server/src/main/java/org/elasticsearch/index/query/CommonTermsQueryBuilder.java [hunk 13]
```diff
@@ -87,6 +99,7 @@
      * Constructs a new common terms query.
      */
     public CommonTermsQueryBuilder(String fieldName, Object text) {
+        deprecationLogger.deprecated(COMMON_TERMS_QUERY_DEPRECATION_MSG);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/CommonTermsQueryBuilder.java]: The whole class is deprecated so you don't need to log the deprecation here. Maybe add a Deprecated annotation to all constructors and you can set the deprecation for the rest handler in the SearchModule. Replacing https://github.com/elastic/elasticsearch/blob/6bba9fc83bc30d29c67e66e8e4c70f4dcc3b7494/server/src/main/java/org/elasticsearch/search/SearchModule.java#L770 with:
````
 registerQuery(new QuerySpec<>(new ParseField(CommonTermsQueryBuilder.NAME).withAllDeprecated(COMMON_TERMS_QUERY_DEPRECATION_MSG),  CommonTermsQueryBuilder::new, CommonTermsQueryBuilder::fromXContent));
````
should work.

File: server/src/main/java/org/elasticsearch/index/query/CommonTermsQueryBuilder.java [hunk 14]
```diff
@@ -102,6 +115,7 @@ public CommonTermsQueryBuilder(String fieldName, Object text) {
      */
     public CommonTermsQueryBuilder(StreamInput in) throws IOException {
         super(in);
+        deprecationLogger.deprecated(COMMON_TERMS_QUERY_DEPRECATION_MSG);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/CommonTermsQueryBuilder.java]: Same here, no need to log here

File: server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryBuilderTests.java [hunk 15]
```diff
@@ -39,6 +40,13 @@
 
 public class CommonTermsQueryBuilderTests extends AbstractQueryTestCase<CommonTermsQueryBuilder> {
 
+    @After
+    public void validateDeprecationMessage() {
+        if (getTestName().equals("testParseFailsWithMultipleFields") == false) {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryBuilderTests.java]: Why is this needed ? Isn't it possible to check the warning in the `testParseFailsWithMultipleFields` test directly ?

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @matriv , I left more comments

File: server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java (hunk 12 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java]: But it's a public field, shouldn't we keep the annotation?

File: server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryBuilderTests.java (hunk 15 above)
Comment (matriv) [reply server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryBuilderTests.java]: With the current approach, every call to the constructor logs a deprecation warning, so all the methods of this class including the ones inherited from the abstract superclass log this warning, apart from this `testParseFailsWithMultipleFields` which throws an exception in advance because of misconfiguration of the query.

**Comment by matriv** [comment]
@jimczi Once again, thx for the comments. They have been addressed and the deprecation messages have been adjusted to accommodate the change.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java [hunk 16]
```diff
@@ -165,7 +164,7 @@ public void testUnknownField() throws IOException {
      * parse exception. Some specific objects do not cause any exception as they can hold arbitrary content; they can be
      * declared by overriding {@link #getObjectsHoldingArbitraryContent()}.
      */
-    public final void testUnknownObjectException() throws IOException {
+    public void testUnknownObjectException() throws IOException {
```
Comment (matriv) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: needs to be overridden to assert the deprecation warning message.

File: server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryBuilderTests.java (hunk 15 above)
Comment (jimczi) [reply server/src/test/java/org/elasticsearch/index/query/CommonTermsQueryBuilderTests.java]: ok got it, thanks

**Review by jimczi — APPROVED** [review]
LGTM
